packages/object/src/Charcoal/Object/RevisionsManager.php

```diff
@@ -0,0 +1,227 @@
+<?php
+
+namespace Charcoal\Object;
+
+use Charcoal\Loader\CollectionLoader;
+use Charcoal\Model\ModelFactoryTrait;
+use Charcoal\Model\ModelInterface;
+use Pimple\Psr11\ServiceLocator;
+use Psr\Container\ContainerExceptionInterface;
+use Psr\Container\NotFoundExceptionInterface;
+use Psr\Log\LoggerAwareTrait;
+
+/**
+ * Revisions Manager
+ *
+ * Service handling revision generation and retrieval.
+ * Can be implemented in an event listener
+ *
+ * Revisions need to act on a ModelInterface,
+ * So to use the revision service, one have to set a ModelInterface beforehand.
+ * Failure to do so will result in an \InvalidArgumentException being thrown.
+ */
+class RevisionsManager
+{
+    use ModelFactoryTrait;
+    use LoggerAwareTrait;
+
+    private RevisionsConfig $revisionConfig;
+    private array $modelRevisionConfig;
+    private ModelInterface $model;
+
+    /**
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
+     */
+    public function __construct(ServiceLocator $locator)
+    {
+        $this->revisionConfig = $locator->get('revisions/config');
+        $this->setModelFactory($locator->get('model/factory'));
+        $this->setLogger($locator->get('logger'));
+    }
+
+    public function __invoke(ModelInterface $model): self
+    {
+        $this->setModel($model);
+
+        return $this;
+    }
+
+    public function generateRevision(): ?ObjectRevisionInterface
+    {
+        $model = $this->getModel();
+
+        // Bail early
+        if (
+            !$this->revisionConfig->isEnabled() ||
+            !$this->revisionEnabled()
+        ) {
+            return null;
+        }
+
+        $revisionProperties = $this->parseRevisionProperties();
+        $revisionObject     = $this->createRevisionObject();
+
+        $revisionObject->createFromObject($model, $revisionProperties);
+
+        if (!empty($revisionObject->getDataDiff())) {
+            $revisionObject->save();
+        }
+
+        return $revisionObject;
+    }
+
+    public function latestRevision(): ObjectRevisionInterface
```

Comment:
```suggestion
    public function getLatestRevision(): ObjectRevisionInterface
```

Reply:
What if this model does not have any revisions, should we still continue the practice of returning an empty model or should we return `null`?

Reply:
done